src/main/scala/rise/core/traversal.scala

```diff
@@ -1,381 +1,205 @@
 package rise.core
 
+import scala.language.implicitConversions
 import arithexpr.arithmetic.NamedVar
 import rise.core.semantics._
 import rise.core.types._
 
-object traversal {
-  sealed abstract class Result[+T](val value: T) {
-    def map[U](f: T => U): Result[U]
-    def mapVisitor(f: Visitor => Visitor): Result[T]
+object Traverse {
```

Comment:
I think that the object name should match the file name (I would say either `traverse.scala`/`object traverse` or `traversal.scala`/`object traversal`)